include/rmm/mr/device/hybrid_memory_resource.hpp

```diff
@@ -0,0 +1,177 @@
+ */
+ template <typename SmallAllocMemoryResource, typename LargeAllocMemoryResource>
+class hybrid_memory_resource : public device_memory_resource {
+ public:
+
+  // The default size used to select between the two allocators.
+  static constexpr std::size_t default_threshold_size = 1 << 22; // 4 MiB
+
+  /**
+   * @brief Construct a new hybrid_memory_resource_object that selects between the two
+   * specified memory_resources based on the `threshold_size`.
+   * 
+   * @param small_mr The memory_resource to use for small allocations.
+   * @param large_mr The memory_resource to use for large allocations.
+   * @param threshold_size Allocations > this size (in bytes) use large_mr. Allocations <= this size
+   *                       use small_mr. Must be a power of two.
+   */
+  hybrid_memory_resource(
+    SmallAllocMemoryResource* small_mr,
+    LargeAllocMemoryResource* large_mr,
+    std::size_t threshold_size = default_threshold_size)
+  : small_mr_{small_mr}, large_mr_{large_mr}, threshold_size_{threshold_size} {
+    RMM_EXPECTS(rmm::detail::is_pow2(threshold_size), "threshold_size must be a power of 2.");
+  }
+
+  /**
+   * @brief Destroy the hybrid-memory_resource object.
+   * 
+   * @note since hybrid_memory_resource does not own its upstream memory_resources, this does not
+   *       free any memory.
+   */
+  virtual ~hybrid_memory_resource() = default;
+
+  /**
+   * @brief Query whether the resource supports use of non-null streams for
+   * allocation/deallocation.
+   *
+   * @returns true
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   * 
+   * @return bool true if the resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override { return false; }
+
+  /**
+   * @brief Get the upstream memory_resource used for small allocations.
+   * 
+   * @return SmallAllocMemoryResource* the upstream resource used for small allocations.
+   */
+  SmallAllocMemoryResource* get_small_mr() { return small_mr_; }
+  /**
+   * @brief Get the upstream memory_resource used for large allocations.
+   * 
+   * @return LargeAllocMemoryResource* the upstream resource used for large allocations.
+   */
+  LargeAllocMemoryResource* get_large_mr() { return large_mr_; }
+
+ private:
+
+  /**
+   * @brief Get the memory resource to use to allocate the requested size.
+   * 
+   * Returns the small memory_resource if `bytes` is <= the threshold size, or the large
+   * memory_resource otherwise.
+   *
+   * @param bytes Requested allocation size in bytes.
+   * @return rmm::mr::device_memory_resource& memory_resource that can allocate the requested size.
+   */
+  rmm::mr::device_memory_resource* get_resource(std::size_t bytes) {
+    if (bytes <= threshold_size_)
+      return small_mr_;
+    else return large_mr_;
```

Comment:
```suggestion
    return (bytes <= threshold_size) ? small_mr_ : large_mr_;
```

Reply:
Because they are different pointer types, your suggestion fails to compile. To make it work, you have to static_cast both of them, which makes for more code than the original, so I'm inclined to leave it alone.

```
return (bytes <= threshold_size_) ?
  static_cast<rmm::mr::device_memory_resource*>(small_mr_) :
  static_cast<rmm::mr::device_memory_resource*>(large_mr_);
```

Reply:
Ah, yeah, didn't think of that.